### PR TITLE
AMDGPU: Fix handling of -0 in round lowering

### LINKED ISSUE #65629: [AMDGPU] The `__builtin_round` intrinsic does not preserve `-0.0`
The standard states the following about the behavior of the `round` functoin:
> If the implementation supports IEEE floating-point arithmetic (IEC 60559), for the std::round function: 
>   * If num is ±0, it is returned, unmodified 

This is currently not respected by the AMDGPU implementation of `__builtin_round`. See the following reproducer for an example.
```c
#include <assert.h>
#include <math.h>
#include <stdio.h>

int main() {
  double r;
  double in = -0.0;
#pragma omp target map(from : r)
  { r = __builtin_round(in); }
  assert(signbit(r) && "Result is no longer negative");
}
```
```shell
$ clang round.c -fopenmp --offload-arch=gfx1030 -fopenmp-offload-mandatory
$ ./a.out
a.out: round.c:10: int main(): Assertion `signbit(r) && "Result is no longer negative"' failed.
[1]    316479 IOT instruction (core dumped)  ./a.out

$ clang round.c -fopenmp --offload-arch=sm_89 -fopenmp-offload-mandatory
$ ./a.out
```

### PR DESCRIPTION
**Pull request by arsenm**
Move the select of the offset to the copysign input, so we end up with -0 + -0 which preserves the sign.

The GlobalISel output leaves something to be desired, it emits some trivially foldable bit ops and also fails to form BFI.

Fixes #65629

### CODE REVIEW

**Comment by jayfoad** [comment]
> The GlobalISel output [...] fails to form BFI.

That's not visible in any of the codegen tests is it? We should have something that shows the final codegen for globalisel.

Code change looks good.

**Comment by arsenm** [comment]
> > The GlobalISel output [...] fails to form BFI.
> 
> That's not visible in any of the codegen tests is it? We should have something that shows the final codegen for globalisel.
> 
> Code change looks good.

Yes, this one is inconsistently tested and only has the mir legalizer test

File: llvm/test/CodeGen/AMDGPU/fneg-combines.ll [hunk 1]
```diff
@@ -41,6 +42,57 @@ define amdgpu_kernel void @v_fneg_add_f32(ptr addrspace(1) %out, ptr addrspace(1
 ; GCN-NEXT: flat_store_dword v{{\[[0-9]+:[0-9]+\]}}, [[ADD]]
 ; GCN-NEXT: s_waitcnt vmcnt(0)
 define amdgpu_kernel void @v_fneg_add_store_use_add_f32(ptr addrspace(1) %out, ptr addrspace(1) %a.ptr, ptr addrspace(1) %b.ptr) #0 {
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/fneg-combines.ll:44]: Does this test pass? Looks like you should have deleted the manual checks above this line, now that there are autogenerated checks below this line.

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/fneg-combines.ll:44]: This isn't supposed to be here

**Review by jhuber6 — APPROVED** [review]
This LGTM and it seems the CI is happy.

**Pull request merged**